// node-opcua-address-space/dist/src/data_access/add_dataItem_stuff.js
"use strict";
/**
 * @module node-opcua-address-space.DataAccess
 */
// Release 1.02  5  OPC Unified Architecture, Part 8
Object.defineProperty(exports, "__esModule", { value: true });
exports.add_dataItem_stuff = add_dataItem_stuff;
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_variant_1 = require("node-opcua-variant");
const definition_Description = "Definition  is a vendor - specific," + " human readable string that specifies how the value of this  DataItem  is calculated.";
const valuePrecision_Description = "";
function add_dataItem_stuff(variable, options) {
    const addressSpace = variable.addressSpace;
    const namespace = addressSpace.getNamespace(variable.nodeId.namespace);
    if (Object.hasOwn(options, "definition") && options.definition !== undefined) {
        namespace.addVariable({
            browseName: { name: "Definition", namespaceIndex: 0 },
            dataType: "String",
            description: definition_Description,
            minimumSamplingInterval: 0,
            modellingRule: options.modellingRule ? "Mandatory" : undefined,
            propertyOf: variable,
            typeDefinition: "PropertyType",
            value: new node_opcua_variant_1.Variant({ dataType: node_opcua_variant_1.DataType.String, value: options.definition })
        });
    }
    if (Object.hasOwn(options, "valuePrecision") && options.valuePrecision !== undefined) {
        (0, node_opcua_assert_1.assert)(typeof options.valuePrecision === "number");
        namespace.addVariable({
            browseName: { name: "ValuePrecision", namespaceIndex: 0 },
            dataType: "Double",
            description: valuePrecision_Description,
            minimumSamplingInterval: 0,
            modellingRule: options.modellingRule ? "Mandatory" : undefined,
            propertyOf: variable,
            typeDefinition: "PropertyType",
            value: new node_opcua_variant_1.Variant({ dataType: node_opcua_variant_1.DataType.Double, value: options.valuePrecision })
        });
    }
}
