// node-opcua-address-space/dist/src/alarms_and_conditions/ua_exclusive_deviation_alarm_impl.js
"use strict";
/**
 * @module node-opcua-address-space.AlarmsAndConditions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UAExclusiveDeviationAlarmImpl = exports.UAExclusiveDeviationAlarmImplBase = void 0;
const node_opcua_assert_1 = require("node-opcua-assert");
const deviation_alarm_helper_1 = require("./deviation_alarm_helper");
const ua_exclusive_limit_alarm_impl_1 = require("./ua_exclusive_limit_alarm_impl");
const ua_limit_alarm_impl_1 = require("./ua_limit_alarm_impl");
class UAExclusiveDeviationAlarmImplBase extends ua_exclusive_limit_alarm_impl_1.UAExclusiveLimitAlarmImplBase {
    static instantiate(namespace, type, options, data) {
        const addressSpace = namespace.addressSpace;
        const exclusiveDeviationAlarmType = addressSpace.findEventType("ExclusiveDeviationAlarmType");
        /* c8 ignore next */
        if (!exclusiveDeviationAlarmType) {
            throw new Error("cannot find ExclusiveDeviationAlarmType");
        }
        (0, node_opcua_assert_1.assert)(type === exclusiveDeviationAlarmType.browseName.toString());
        const alarm = ua_exclusive_limit_alarm_impl_1.UAExclusiveLimitAlarmImplBase.instantiate(namespace, type, options, data);
        Object.setPrototypeOf(alarm, exports.UAExclusiveDeviationAlarmImpl.prototype);
        (0, node_opcua_assert_1.assert)(alarm instanceof exports.UAExclusiveDeviationAlarmImpl);
        (0, node_opcua_assert_1.assert)(alarm instanceof ua_exclusive_limit_alarm_impl_1.UAExclusiveLimitAlarmImpl);
        (0, node_opcua_assert_1.assert)(alarm instanceof ua_limit_alarm_impl_1.UALimitAlarmImpl);
        alarm._install_setpoint(options);
        return alarm;
    }
    get $16() {
        return this;
    }
    getSetpointNodeNode() {
        return deviation_alarm_helper_1.DeviationAlarmHelper_getSetpointNodeNode.call(this.$16);
    }
    getSetpointValue() {
        return deviation_alarm_helper_1.DeviationAlarmHelper_getSetpointValue.call(this.$16);
    }
    _onSetpointDataValueChange(dataValue) {
        deviation_alarm_helper_1.DeviationAlarmHelper_onSetpointDataValueChange.call(this.$16, dataValue);
    }
    _install_setpoint(options) {
        deviation_alarm_helper_1.DeviationAlarmHelper_install_setpoint.call(this.$16, options);
    }
    _setStateBasedOnInputValue(value) {
        const setpointValue = this.getSetpointValue();
        if (setpointValue === null) {
            return;
        }
        (0, node_opcua_assert_1.assert)(Number.isFinite(setpointValue));
        // call base class implementation
        ua_exclusive_limit_alarm_impl_1.UAExclusiveLimitAlarmImpl.prototype._setStateBasedOnInputValue.call(this, value - setpointValue);
    }
}
exports.UAExclusiveDeviationAlarmImplBase = UAExclusiveDeviationAlarmImplBase;
exports.UAExclusiveDeviationAlarmImpl = UAExclusiveDeviationAlarmImplBase;
